presto-main/src/main/java/com/facebook/presto/execution/scheduler/NetworkLocationCache.java

```diff
@@ -0,0 +1,88 @@
+package com.facebook.presto.execution.scheduler;
+
+import com.facebook.presto.spi.HostAddress;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+
+import static com.google.common.cache.CacheLoader.asyncReloading;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static java.util.concurrent.TimeUnit.DAYS;
+import static java.util.concurrent.TimeUnit.HOURS;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.MINUTES;
+
+public class NetworkLocationCache
+{
+    private static final Duration NEGATIVE_CACHE_DURATION = new Duration(10, MINUTES);
+
+    @SuppressWarnings("ThrowableInstanceNeverThrown")
+    private static final RuntimeException NEGATIVE_CACHE_EXCEPTION = new RuntimeException("value negatively cached");
+
+    private static final Logger log = Logger.get(NetworkLocationCache.class);
+
+    private final ExecutorService executor = newCachedThreadPool(daemonThreadsNamed("network-location-%s"));
+    private final NetworkTopology networkTopology;
+    private final LoadingCache<HostAddress, NetworkLocation> cache;
+    private final Cache<HostAddress, Boolean> negativeCache;
+
+    public NetworkLocationCache(NetworkTopology networkTopology)
+    {
+        this.networkTopology = requireNonNull(networkTopology, "networkTopology is null");
+
+        this.cache = CacheBuilder.newBuilder()
+                .expireAfterWrite(1, DAYS)
+                .refreshAfterWrite(12, HOURS)
+                .build(asyncReloading(new CacheLoader<HostAddress, NetworkLocation>()
+                {
+                    @Override
+                    public NetworkLocation load(HostAddress host)
+                            throws Exception
+                    {
+                        return locate(host);
+                    }
+                }, executor));
+
+        this.negativeCache = CacheBuilder.newBuilder()
+                .expireAfterWrite(NEGATIVE_CACHE_DURATION.toMillis(), MILLISECONDS)
+                .build();
+    }
+
+    public void stop()
+    {
+        executor.shutdownNow();
+    }
+
+    public Optional<NetworkLocation> get(HostAddress host)
+    {
+        NetworkLocation location = cache.getIfPresent(host);
+        if (location == null) {
+            cache.refresh(host);
+        }
+        return Optional.ofNullable(location);
+    }
+
+    private NetworkLocation locate(HostAddress host)
+    {
+        if (negativeCache.getIfPresent(host) != null) {
```

Comment:
I think it would be better to put this check on line 67 in the `if (location == null)`. Otherwise we could end up throwing a lot of exceptions in the case where everything is negatively cached, since it will always go through the `cache.refresh()` code path